File: tests/demobase.py

import sys
from operator import sub

import OpenGL.GL as gl
import OpenGL.GLU as glu
import OpenGL.GLUT as glut

from orbitcamera import OrbitCamera

class CameraState:
    NONE = 0
    ROTATION = 1
    TRANSLATION = 2
    ZOOM = 3

class DemoBase(object):
    
    def __init__(self, width, height):
        self.window = None
        self.window_width, self.window_height = width, height
        self.key_state = [False]*256
        self.mouse_position = None
        self.camera = OrbitCamera()
        self.camera_state = CameraState.NONE
    
    def run(self):
        glut.glutInit()
        glut.glutInitDisplayMode(glut.GLUT_RGBA | glut.GLUT_DOUBLE
                                    | glut.GLUT_ALPHA | glut.GLUT_DEPTH)
        glut.glutInitWindowSize(self.window_width, self.window_height)
        
        self.window = glut.glutCreateWindow("Point cloud test")
        
        glut.glutDisplayFunc(self.display_base)
        glut.glutIdleFunc(self.display_base)
        glut.glutReshapeFunc(self.resize_event)
        glut.glutKeyboardFunc(self.keyboard_press_event)
        glut.glutKeyboardUpFunc(self.keyboard_up_event)
        glut.glutMouseFunc(self.mouse_button_event)
        glut.glutMotionFunc(self.mouse_moved_event)
        
        self.init_gl(self.window_width, self.window_height)
        
        glut.glutMainLoop()
    
    def init_gl(self, width, height):
        gl.glClearColor(0.2, 0.2, 0.2, 0.0)
        gl.glEnable(gl.GL_DEPTH_TEST)
        
        self.resize_event(width, height)
    
    def resize_event(self, width, height):
        gl.glViewport(0, 0, width, height)
        gl.glMatrixMode(gl.GL_PROJECTION)
        gl.glLoadIdentity()
        glu.gluPerspective(60, width/float(height), 50.0, 8000.0)
        gl.glMatrixMode(gl.GL_MODELVIEW)
        
        self.window_width = width
        self.window_height = height
    
    def display_base(self):
        gl.glClear(gl.GL_COLOR_BUFFER_BIT | gl.GL_DEPTH_BUFFER_BIT)
        
        # Prepare the model-view matrix.
        gl.glMatrixMode(gl.GL_MODELVIEW)
        gl.glLoadIdentity()
        glu.gluLookAt(*self.camera.get_glulookat_parameters())
        
        self.display()
        
        glut.glutSwapBuffers()
    
    def keyboard_press_event(self, key, x, y):
        
        if key == chr(27):
            glut.glutDestroyWindow(self.window)
            sys.exit(0)
        
        self.key_state[ord(key)] = True
    
    def keyboard_up_event(self, key, x, y):
        self.key_state[ord(key)] = False
    
    def mouse_button_event(self, button, state, x, y):
        self.mouse_position = (x,y)
        
        if state == glut.GLUT_UP:
            self.camera_state = CameraState.NONE
        elif button == glut.GLUT_LEFT_BUTTON:
            self.camera_state = CameraState.ROTATION
        elif button == glut.GLUT_RIGHT_BUTTON:
            self.camera_state = CameraState.ZOOM
        elif button == glut.GLUT_MIDDLE_BUTTON:
            self.camera_state = CameraState.TRANSLATION
    
    def mouse_moved_event(self, x, y):
        offset = map(sub, (x,y), self.mouse_position)
        
        if self.camera_state == CameraState.ROTATION:
            self.camera.rotate(*[i/100.0 for i in offset])
        elif self.camera_state == CameraState.TRANSLATION:
            self.camera.translate(offset[0]*3, -offset[1]*3)
        elif self.camera_state == CameraState.ZOOM:
            self.camera.zoom(offset[1])
        
        self.mouse_position = (x,y)
    

File: src/DemoBase.cpp

#include "DemoBase.h"

#include "OrbitCamera.h"

#include <algorithm>
#include <stdexcept>
#include <cstdlib>

#if defined(__APPLE__) || defined(MACOSX)
#include <OpenGL/gl.h>
#include <GLUT/glut.h>
#else
#include <GL/gl.h>
#include <GL/glut.h>
#endif

DemoBase* DemoBase::instance = 0;

// Trampoline functions.
inline void DemoBase::display_base()
{
    DemoBase::instance->displayBase();
}

inline void DemoBase::resize_event(int width, int height)
{
    DemoBase::instance->resizeEvent(width, height);
}

inline void DemoBase::keyboard_press_event(unsigned char key, int x, int y)
{
    DemoBase::instance->keyboardPressEvent(key, x, y);
}

inline void DemoBase::keyboard_up_event(unsigned char key, int x, int y)
{
    DemoBase::instance->keyboardUpEvent(key, x, y);
}

inline void DemoBase::mouse_button_event(int button, int state, int x, int y)
{
    DemoBase::instance->mouseButtonEvent(button, state, x, y);
}

inline void DemoBase::mouse_moved_event(int x, int y)
{
    DemoBase::instance->mouseMovedEvent(x, y);
}

DemoBase::DemoBase(int width, int height)
    : mCameraState(NONE), mWindow(0), mWidth(width), mHeight(height)
{
    if(instance != 0)
        throw std::runtime_error("There is already an instance of DemoBase.");
    
    instance = this;
    mCamera = new OrbitCamera();
    std::fill(mKeyState, mKeyState+256, false);
    std::fill(mMousePosition, mMousePosition+2, 0);
}

DemoBase::~DemoBase()
{
    delete mCamera;
}

void DemoBase::run(int* argcp, char** argv)
{
    glutInit(argcp, argv);
    glutInitDisplayMode(GLUT_RGBA | GLUT_DOUBLE | GLUT_ALPHA | GLUT_DEPTH);
    glutInitWindowSize(mWidth, mHeight);
    
    mWindow = glutCreateWindow("FreenectFusion");
    
    glutDisplayFunc(DemoBase::display_base);
    glutIdleFunc(DemoBase::display_base);
    glutReshapeFunc(DemoBase::resize_event);
    glutKeyboardFunc(DemoBase::keyboard_press_event);
    glutKeyboardUpFunc(DemoBase::keyboard_up_event);
    glutMouseFunc(DemoBase::mouse_button_event);
    glutMotionFunc(DemoBase::mouse_moved_event);
    
    initGl(mWindow, mHeight);
    
    glutMainLoop();
}

void DemoBase::initGl(int width, int height)
{
    glClearColor(0.2, 0.2, 0.2, 0.0);
    glEnable(GL_DEPTH_TEST);
    resizeEvent(width, height);
}

void DemoBase::resizeEvent(int width, int height)
{
    double aspect = static_cast<double>(width)/static_cast<double>(height);
    glViewport(0, 0, width, height);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(60, aspect, 50.0, 8000.0);
    glMatrixMode(GL_MODELVIEW);
    
    mWidth = width;
    mHeight = height;
}

void DemoBase::displayBase()
{
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    double params[9];
    
    // Prepare the model-view matrix.
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    mCamera->getGluLookAtParameters(params);
    gluLookAt(params[0], params[1], params[2], 
              params[3], params[4], params[5],
              params[6], params[7], params[8]);
    display();
    glutSwapBuffers();      
}

void DemoBase::keyboardPressEvent(unsigned char key, int x, int y)
{
    if(key == 27)
    {
        glutDestroyWindow(mWindow);
        std::exit(0);
    }
    
    mKeyState[key] = true;
}

void DemoBase::keyboardUpEvent(unsigned char key, int x, int y)
{
    mKeyState[key] = false;
}

void DemoBase::mouseButtonEvent(int button, int state, int x, int y)
{
    mMousePosition[0] = x; mMousePosition[1] = y;
    
    if(state == GLUT_UP)
    {
        mCameraState = NONE;
        return;
    }
    
    switch(button)
    {
    case GLUT_LEFT_BUTTON:
        mCameraState = ROTATION;
        break;
    case GLUT_RIGHT_BUTTON:
        mCameraState = ZOOM;
        break;
    case GLUT_MIDDLE_BUTTON:
        mCameraState = TRANSLATION;
        break;
    default:;
    }
}

void DemoBase::mouseMovedEvent(int x, int y)
{
    double offx = x-mMousePosition[0];
    double offy = y-mMousePosition[1];
    
    switch(mCameraState)
    {
    case ROTATION:
        mCamera->rotate(offx/100.0, offy/100.0);
        break;
    case TRANSLATION:
        mCamera->translate(offx*3, -offy*3);
        break;
    case ZOOM:
        mCamera->zoom(offy);
        break;
    default:;
    }
    
    mMousePosition[0] = x; mMousePosition[1] = y;
}
